### PR TITLE
Fixed crasher 28707

### PR DESCRIPTION
**Pull request by hartbit**
The `@_versioned` accessibility check is done against the formal access. But the diagnostics is printed with `getAccessForDiagnostics`. In cases where the effective access is internal, the diagnostics will crash. Example:

```swift
internal class A {
    @_versioned public var a: Int = 0
}
```

This change was introduced in https://github.com/apple/swift/commit/f2d159940cdeab6333f154a9f25eac551855ada5 by @slavapestov when fixing `@_fixed_layout`.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by slavapestov** [comment]
@hartbit Can you add some more test cases to test/attr/attr_versioned.swift?

**Comment by hartbit** [comment]
@slavapestov I added some more tests but I'm a bit confused. Can you answer a few questions?

```
internal class internalClass {
  @_versioned public func publicVersioned() {}
  // formal access: public
  // effective access: internal
  // This currently generates an error. But should it? The decl is effectively internal.
}

fileprivate class filePrivateClass {
  @_versioned internal func internalVersioned() {}
  // formal access: internal
  // effective access: fileprivate
  // This currently does not generate an error. But it should, right? The decl is effectively fileprivate.
}
```

**Comment by slavapestov** [comment]
We allow decls to be more visible than their surrounding context. In this case we simply set the visibility of the decl to be that of the surrounding context. So I think these two examples are not errors, if we want to treat @_versioned the same as other visibility modifiers.

File: validation-test/compiler_crashers_fixed/28707-false-encountered-error-in-diagnostic-text.swift [hunk 1]
```diff
@@ -6,7 +6,7 @@
 // See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 
 // REQUIRES: asserts
```
Comment (slavapestov) [inline validation-test/compiler_crashers_fixed/28707-false-encountered-error-in-diagnostic-text.swift]: Please remove the REQUIRES line -- once a crasher is fixed, it no longer requires asserts to be on to pass

**Comment by slavapestov** [comment]
The new tests look fine. I'm OK with merging this if you are.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: validation-test/compiler_crashers_fixed/28707-false-encountered-error-in-diagnostic-text.swift (hunk 1 above)
Comment (hartbit) [inline validation-test/compiler_crashers_fixed/28707-false-encountered-error-in-diagnostic-text.swift]: Done!

**Comment by hartbit** [comment]
@slavapestov I'm ok for merging! -1 crash 🎉

**Comment by CodaFi** [comment]
@swift-ci please smoke test and merge

**Comment by CodaFi** [comment]
@swift-ci please smoke test and merge

**Pull request merged**